### DISCUSSION TITLE
If the same space is created twice, are they exactly the same?

### DISCUSSION THREAD

**Discussion by jrmaddison** [firedrakeproject/firedrake#3357] [Firedrake support] [answered]
I see that if I create the same space twice e.g. `space0 = FunctionSpace(mesh, "Lagrange", 1)` and `space1 = FunctionSpace(mesh, "Lagrange", 1)` that they share some data which is cached on the mesh. Is there a reason `space0` and `space1` are not the same object -- i.e. is there anything actually different about them? Or can I use them interchangeably throughout my code?

**Comment by wence-** [comment, marked as answer]
> I see that if I create the same space twice e.g. `space0 = FunctionSpace(mesh, "Lagrange", 1)` and `space1 = FunctionSpace(mesh, "Lagrange", 1)` that they share some data which is cached on the mesh. Is there a reason `space0` and `space1` are not the same object

I believe when we implemented singleton patterns for function spaces, we decided just to cache the "heavy" data, so that `FunctionSpace` itself is a lightweight thing that is "cheap" to make, as long as one with the same dof layout has already been created. Part of the rationale is that the expensive-to-construct things are actually less specific than the cache key you would naively build using the "obvious" information that is available from the mesh and ufl element. e.g. a scalar Lagrange element has the same `cell_node_map` irrespective of the immersion dimension of the mesh.

> -- i.e. is there anything actually different about them? Or can I use them interchangeably throughout my code?

If you can't use them interchangeably, that is a bug.
